src/typecheck/eq.rs

```diff
@@ -43,24 +43,61 @@
 //! times. For anything more complex, we return false.
 
 use super::*;
-use crate::eval;
-use crate::term::UnaryOp;
+use crate::{eval, term::UnaryOp};
 
 /// The maximal number of variable links we want to unfold before abandoning the check. It should
 /// stay low, but has been fixed arbitrarily: feel fee to increase reasonably if it turns out
 /// legitimate type equalities between simple contracts are unduly rejected in practice.
 pub const MAX_GAS: u8 = 8;
 
+/// Abstract over the term environment, which is represented differently in the typechecker and
+/// during evaluation.
+///
+/// The evaluation environment holds [`crate::eval::lazy::Thunks`], which are `Rc<RefCell<_>>`
+/// under the hood, while the term environment used during typechecking is just maps identifiers to
+/// a pair `(RichTerm, Environment)`. To have an interface that works with both,
+/// `TermEnvironment::get` has to take a closure representing the continuation of the task to do
+/// with the result instead of merely returning it.
+pub trait TermEnvironment: Clone {
+    fn get_then<F, T>(&self, id: &Ident, f: F) -> T
+    where
+        F: FnOnce(Option<(&RichTerm, &Self)>) -> T;
+}
+
+/// A simple term environment, as a mapping from identifiers to a tuple of a term and an
+/// environment (i.e. a closure), sufficient for the needs of typechecking.
 #[derive(PartialEq, Clone, Debug)]
-pub struct TermEnvironment(pub GenericEnvironment<Ident, (RichTerm, TermEnvironment)>);
+pub struct SimpleTermEnvironment(pub GenericEnvironment<Ident, (RichTerm, SimpleTermEnvironment)>);
 
-impl TermEnvironment {
+impl SimpleTermEnvironment {
     pub fn new() -> Self {
-        TermEnvironment(GenericEnvironment::new())
+        SimpleTermEnvironment(GenericEnvironment::new())
+    }
+}
+
+impl TermEnvironment for SimpleTermEnvironment {
+    fn get_then<F, T>(&self, id: &Ident, f: F) -> T
+    where
+        F: FnOnce(Option<(&RichTerm, &SimpleTermEnvironment)>) -> T,
+    {
+        f(self.0.get(id).map(|(rt, env)| (rt, env)))
+    }
+}
+
+impl TermEnvironment for eval::Environment {
+    fn get_then<F, T>(&self, id: &Ident, f: F) -> T
+    where
+        F: FnOnce(Option<(&RichTerm, &eval::Environment)>) -> T,
+    {
+        if let Some(closure_ref) = self.get(id).map(eval::lazy::Thunk::borrow) {
+            f(Some((&closure_ref.body, &closure_ref.env)))
+        } else {
+            f(None)
+        }
```

Comment:
I don't have a personal opinion but I think clippy favors `if-let` for single matches: https://rust-lang.github.io/rust-clippy/master/#single_match. I would tend to follow this trend, if only because it's standard if enforced by clippy.

Reply:
That's not what the clippy lint is about, it's for when you pattern match on only one case and ignore others. Here you still have to do something when the case is None.

Reply:
Ok, fair enough, clippy doesn't warn. Although this is really nitpicking, both forms are valid and idiomatic IMHO.